### PR TITLE
Add keys for back and forward through history of visited rooms

### PR DESCRIPTION
**Pull request by MRAAGH**
Use case: when you are active in many rooms at the same time (3 rooms, 4 rooms, N rooms), the history of recently visited rooms contains exactly what you need to quickly switch to. These bindings for going back and forwards through the history list enable quick switching while maintaining list order.

Inspiration came from Element which has this functionality on Alt+left and Alt+right (perhaps accidentally, because it is a browser).

I picked the bindings `Ctrl+H` and `Ctrl+Y` because they are easily reachable, adjacent, and `H` is for "History".

This does not change the functionality of `last_page` or any other room switching. However, if one does that while not at the end of history, a portion of the history will be discarded. Analogy: in the browser, go a few pages back, then click a different link.

### CODE REVIEW

File: src/gui/PageLoader.qml [hunk 1]
```diff
@@ -61,6 +71,22 @@ HLoader {
         return true
     }
 
+    function moveThroughVisitHistory(relativeMovement=1) {
```
Comment (mirukana) [inline src/gui/PageLoader.qml:74]: We can shorten the name to `moveThroughHistory`. I think this can replace the `showPrevious` function too.

File: src/gui/PageLoader.qml [hunk 2]
```diff
@@ -61,6 +71,22 @@ HLoader {
...
+
+        // not allowed to go beyond oldest entry in history
+        if (historyPosition + relativeMovement >= history.length) return false
+
+        // not allowed to go beyond newest entry in history
+        if (historyPosition + relativeMovement < 0) return false
```
Comment (mirukana) [inline src/gui/PageLoader.qml:80]: It would be more useful to wrap around here rather than just returning

**Review by mirukana — CHANGES_REQUESTED** [review]
Very useful, thanks! Some concerns besides the comments below: I want to have a better ctrl+(shift+)tab in the future, similar to desktop alt+tab or firefox classic ctrl+tab.

Besides showing the history as a visual list while the keys are pressed and maybe taking advantage of synced breadcrumbs (#180); the navigation would work like the feature added with this PR, except the first page to be offered as a choice to move to would always be the one the user just left, to allow for easy back-and-forth switching.

Basically, the currrent ctrl+tab and this new ctrl+h would be combined into one feature, do you see any issue with this and need a pure back/forward to remain?

File: src/config/settings.py [hunk 3]
```diff
@@ -253,6 +253,11 @@ def alt_or_cmd(self) -> str:
     # Switch to the last opened page/chat, similar to Alt+Tab on most desktops.
     last_page = ["Ctrl+Tab"]
 
+    # Go throgh history of opened chats,
+    # similar to the "page back" and "page forward" keys in web browsers
+    visit_history_back = ["Ctrl+H"]
+    visit_history_forward = ["Ctrl+Y"]
```
[the remark this answers is not in the document]
Comment (MRAAGH) [reply src/config/settings.py:259]: Ah, see I forgot that `Z` and `Y` are swapped on a lot of keyboard layouts, so this is indeed not a good default binding (I'll use it myself anyway). `Ctrl+Shift+H` is better. In fact, if we want to make it intuitive, we could even make it `Alt+Left` and `Alt+Right` like in browsers; and maybe even mouse button 4 and mouse button 5.

About the setting names; I was trying to show that it's about the history of visited rooms, rather than history of received messages etc. but I suppose the comment also does it justice, so I'll shorten the names as you suggested.

[the remark this answers is not in the document]
Comment (mirukana) [reply src/config/settings.py:259]: Alt+Left/Right are already used for horizontal scrolling (e.g. in the image viewer), but extra mouse controls sound good. This can be done using [`TapHandler`s](https://doc.qt.io/qt-5/qml-qtquick-taphandler.html) with `acceptedButtons` set to one of the [supported buttons](https://doc.qt.io/qt-5/qt.html#MouseButton-enum).

How about `earlier_page` and `later_page` for the settings name? That's consistent with the `last_page` too

**Comment by MRAAGH** [comment]
> Besides showing the history as a visual list while the keys are pressed and maybe taking advantage of synced breadcrumbs (#180); the navigation would work like the feature added with this PR, except the first page to be offered as a choice to move to would always be the one the user just left, to allow for easy back-and-forth switching.
Basically, the currrent ctrl+tab and this new ctrl+h would be combined into one feature, do you see any issue with this and need a pure back/forward to remain?

Bringing the rooms to the top when they are switched to keeps the recently visited rooms accessible with the fewest keypresses. However the drawback is that the order is changing so the user can't memorize it. Instead, it requires visual feedback, displaying the options (what you'll get after 1 keypress, 2, 3 etc.). The human needs to recognize the displayed icons/names and decide which is the one they want to switch to. (and if they pick incorrectly, the rooms get reordered again)

It comes down to this: pressing a key 4 times where you already knew ahead of time that you have to press it 4 times is *faster* than reading information on the screen, interpreting it, then doing the appropriate action. I think that for power users, software that does not require visual feedback is more efficient.

File: src/gui/PageLoader.qml (hunk 1 above)
Comment (MRAAGH) [reply src/gui/PageLoader.qml:74]: `moveThroughHistory` and `showPrevious` have different checks, one acknowledges `historyPosition` while the other ignores it (because it will be reset). I could merge these two functions and add an argument for picking between one and the other behavior, but I think that would make the code less neatly organized. What do you think?

File: src/gui/PageLoader.qml (hunk 2 above)
Comment (MRAAGH) [reply src/gui/PageLoader.qml:80]: I should call `show` for switching to the room that's currently shown already?

File: src/gui/PageLoader.qml (hunk 1 above)
Comment (mirukana) [reply src/gui/PageLoader.qml]: I'm bothered by how the two functions end with the same 4 lines, but you can keep it that way for now I suppose

File: src/gui/PageLoader.qml (hunk 2 above)
Comment (mirukana) [reply src/gui/PageLoader.qml:80]: - If there's no entries in the history, just return
- if user goes back while already being at index 0, the new index should be the last history entry
- if user goes forward while being at last index, new index should be 0

Comment (MRAAGH) [reply src/gui/PageLoader.qml:80]: It comes down to personal preference. For me, I'd much rather have the button do nothing, as an indication that I'm at the end of the list. So I added a setting for this. See latest commit

File: src/gui/PageLoader.qml (hunk 1 above)
Comment (MRAAGH) [reply src/gui/PageLoader.qml]: I grouped that part into a new function that gets used in 2 places. I couldn't think of anything better.

**Comment by mirukana** [comment]
Last thought before merging (sorry for the delay): when using Ctrl+H for back, I keep hitting Ctrl+L thinking it would be forward, since HJKL/arrows are used for many other directional/navigation bindings.

I realized then we could have Ctrl+H/Left and Ctrl+L/Right as bindings, this feels more natural to me (no moving off the homerow) and would give you a Left/Right similar to browsers. We can switch the default bind of `Keys.Messages.clear_all` to Ctrl+Shift+L, as this isn't a frequently used or that useful of a function. Does that sounds good?

**Comment by MRAAGH** [comment]
> Last thought before merging (sorry for the delay): when using Ctrl+H for back, I keep hitting Ctrl+L thinking it would be forward, since HJKL/arrows are used for many other directional/navigation bindings.
> 
> I realized then we could have Ctrl+H/Left and Ctrl+L/Right as bindings, this feels more natural to me (no moving off the homerow) and would give you a Left/Right similar to browsers. We can switch the default bind of `Keys.Messages.clear_all` to Ctrl+Shift+L, as this isn't a frequently used or that useful of a function. Does that sounds good?

This sounds perfect; the mnemonic for Ctrl+H could be "history" and mnemonic for Ctrl+L could be "later"! And yes, clearing all messages doesn't seem like a common operation, and doesn't need to be on an accessible binding, especially since it could be considered a destructive action.

About the mouse bindings (mb4 and mb5), I tried adding them but they were only being acknowledged if the mouse was on specific places on the screen. Haven't figured out how to do this properly. I might try again later. It's fine to merge this without them.
